keps/0021-kudo-upgrade.md

```diff
@@ -0,0 +1,271 @@
+            * [Failure cases](#failure-cases)
+      * [Graduation Criteria TODO](#graduation-criteria-todo)
+      * [Drawbacks TODO [optional]](#drawbacks-todo-optional)
+      * [Alternatives](#alternatives)
+      * [Infrastructure Needed](#infrastructure-needed)
+      * [Resources](#resources)
+      * [Implementation History](#implementation-history)
+
+<!-- Added by: aneumann, at: Mon Dec  2 11:34:32 CET 2019 -->
+
+<!--te-->
+
+## Summary
+
+TBD
+
+## Motivation
+
+We strive for quick and regular updates of KUDO. We need a process for upgrading all the moving parts of KUDO, and how 
+the different parts interact, and what kind of compatibility we want to provide.
+
+### Goals
+
+- How updates of KUDO are executed
+  - Updates of CRDs
+  - Updates of the KUDO manager
+  - Updates of prerequisites (Namespaces, RoleBindings, ServiceAccounts, etc.)
+- Interoperability
+  - How and if multiple versions of CRDs are maintained
+  - How a version of KUDO CLI work with older and newer CRD versions
+- How to handle operators that are not supported by a new KUDO version
+
+### Non-Goals
+
+- Updates/Upgrades of operators and operator versions itself
+- Versioning of KUDO itself
+- Multi-Tenancy, i.e. running multiple KUDO installations on the same cluster
+
+
+### Current State
+
+At the moment, KUDO does not provide any migration capabilities and needs a clean installation to use a new version.
+
+## Open Questions
+- Lowest supported K8s version
+- Do we want to support downgrades? 
+- Split versions between KUDO manager and KUDO CLI?
+
+## Proposal
+
+### KUDO Prerequisites
+Expected update frequency: Low  
+Versioned: No, but closely tied to KUDO manager
+
+The KUDO manager has a set of prerequisites that need to be present for the manager to run successfully. They are
+the least likely to change, but probably the most specific. If we make changes here, we need to implement custom
+migration code.
+
+- Current prereqs
+  - Namespace
+  - Service Account
+  - Role Bindings
+  - Secret
+
+#### Proposal for update process 
+Integrated into `kudo init --upgrade`
+
+Write specific migration code that targets a KUDO manager version range and executes manual migration steps. 
+
+Each migration should have a validate-step that checks if the migration is possible.
+
+**Alternative update process**
+- Can we just delete them all and reinstall them? Probably not
+
+### KUDO Manager
+Expected update frequency: High  
+Versioned: Yes
+
+The KUDO Manager is defined by an image version in a deployment set. To update, the deployment must be updated. The 
+manager is closely tied to the CRDs, but not to the CLI. When CRDs are updated, the Manager will most likely also
+need to be updated. 
+
+#### Proposal for update process
+Integrated into `kudo init --upgrade`
+
+- Use semantic versioning for the manager binary
+  - ToBeDiscussed: Use different versioning for Manager and CLI?
+- The manager needs to be shut down before updating the CRDs
```

Comment:
does it? why? also are we talking about updating crd of the same version or updating to newer version or both?

Reply:
Well - this assumes we're not using the conversion web hook, and therefore can't serve different versions of a CR and need to migrate them manually. In that case, having a manager running that tries to read or modify the old version of CRs would be problematic.

Regarding updating CRDs with the same version - Is that something that is commonly used? I can imagine updating a CRD with new fields, assuming it's backwards compatible and defaults are provided, but in which case is the version of a CRD updated then? Only in case of breaking changes?

Reply:
I just read the K8s documentation on API changes. (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md) Explains a lot. I'll update the KEP.